src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -990,275 +1000,170 @@ private static Expression AddResponseWritingToMethodCall(Expression methodCall,
 
-            // Looping over arrays is faster
-            var binders = factoryContext.ParameterBinders.ToArray();
-            var count = binders.Length;
+            var (_, binder) = factoryContext.AsyncParameters[0];
 
             return async (target, httpContext) =>
             {
-                // Run these first so that they can potentially read and rewind the body
-                var boundValues = new object?[count];
-
-                for (var i = 0; i < count; i++)
-                {
-                    boundValues[i] = await binders[i](httpContext);
-                }
+                var boundValue = await binder(httpContext);
 
-                var (bodyValue, successful) = await TryReadBodyAsync(
-                    httpContext,
-                    bodyType,
-                    parameterTypeName,
-                    parameterName,
-                    factoryContext.AllowEmptyRequestBody,
-                    factoryContext.ThrowOnBadRequest);
-
-                if (!successful)
+                if (ReferenceEquals(boundValue, FailedBodyReadAlreadyHandledSentinel))
                 {
+                    // A default parameter binder has already logged the failed body read. We're done.
                     return;
                 }
 
-                await continuation(target, httpContext, bodyValue, boundValues);
+                await continuation(target, httpContext, boundValue);
             };
         }
         else
         {
-            // We need to generate the code for reading from the body before calling into the delegate
-            var continuation = Expression.Lambda<Func<object?, HttpContext, object?, Task>>(
-            responseWritingMethodCall, TargetExpr, HttpContextExpr, BodyValueExpr).Compile();
+            // We need to generate the code for reading from the custom binders calling into the delegate
+            var continuation = Expression.Lambda<Func<object?, HttpContext, object?[], Task>>(
+                responseWritingMethodCall, TargetExpr, HttpContextExpr, AsyncValuesArrayExpr).Compile();
+
+            var binders = new Func<HttpContext, ValueTask<object?>>[factoryContext.AsyncParameters.Count];
+            for (var i = 0; i < binders.Length; i++)
+            {
+                (_, binders[i]) = factoryContext.AsyncParameters[i];
+            }
+            var count = binders.Length;
 
             return async (target, httpContext) =>
             {
-                var (bodyValue, successful) = await TryReadBodyAsync(
-                    httpContext,
-                    bodyType,
-                    parameterTypeName,
-                    parameterName,
-                    factoryContext.AllowEmptyRequestBody,
-                    factoryContext.ThrowOnBadRequest);
-
-                if (!successful)
+                var boundValues = new object?[count];
+
+                // Looping over arrays is faster
```

Comment:
Comment was more for when we called `.ToArray()` on the binders collection, can remove